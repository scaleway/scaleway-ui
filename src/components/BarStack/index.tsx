import styled from '@emotion/styled'
import { darken, lighten, transparentize } from 'polished'
import React, { MouseEventHandler, ReactNode, useMemo } from 'react'
import Tooltip from '../Tooltip'

export type BarProps = {
  /**
   * Unique id of the bar
   */
  id: string
  /**
   * The value of the bar
   */
  value: number
  /**
   * Text to display inside the bar
   */
  text?: string
  onClick?: MouseEventHandler<HTMLDivElement>
  onDoubleClick?: MouseEventHandler<HTMLDivElement>
  onMouseDown?: MouseEventHandler<HTMLDivElement>
  onMouseUp?: MouseEventHandler<HTMLDivElement>
  onMouseEnter?: MouseEventHandler<HTMLDivElement>
  onMouseLeave?: MouseEventHandler<HTMLDivElement>
  /**
   * A tooltip to display when hovering the bar
   */
  tooltip?: ReactNode
}

export interface BarStackProps {
  data: Array<BarProps>
  total?: number
}

const StyledBarWrapper = styled.div`
  width: 0px;
  transition: width 500ms;
  background-color: ${({ theme }) => theme.colorsDeprecated.white};
`

const StyledBar = styled.div`
  height: 50px;
  font-weight: 600;
  color: ${({ theme }) => theme.colorsDeprecated.white};
  font-size: 14px;
  display: flex;
  align-items: center;
  padding: ${({ theme }) => theme.space[1]};
  width: 100%;
  white-space: nowrap;
  overflow: hidden;
  text-shadow: -1px 0
      ${({ theme }) => transparentize(0.7, theme.colorsDeprecated.black)},
    0 1px ${({ theme }) => transparentize(0.7, theme.colorsDeprecated.black)},
    1px 0 ${({ theme }) => transparentize(0.7, theme.colorsDeprecated.black)},
    0 -1px ${({ theme }) => transparentize(0.7, theme.colorsDeprecated.black)};
`

const StyledContainer = styled.div`
  width: 100%;
  display: flex;
  background-color: ${({ theme }) => theme.colorsDeprecated.gray100};
  border-radius: ${({ theme }) => theme.radii.default};
  box-shadow: inset 0px 0px 0px 1px
    ${({ theme }) => theme.colorsDeprecated.gray300};
  overflow: hidden;

  ${StyledBarWrapper}:nth-child(5n+1) {
    ${({ theme }) => `background: linear-gradient(-45deg, ${transparentize(
      0.9,
      theme.colorsDeprecated.white,
    )} 25%,
      ${theme.colorsDeprecated.primary} 25%, ${
      theme.colorsDeprecated.primary
    } 50%,
      ${transparentize(
        0.9,
        theme.colorsDeprecated.white,
      )} 50%, ${transparentize(0.9, theme.colorsDeprecated.white)} 75%, ${
      theme.colorsDeprecated.primary
    }
       75%);`}
    background-size: 30px 30px;
    background-color: ${({ theme }) => theme.colorsDeprecated.primary};
  }

  ${StyledBarWrapper}:nth-child(5n+2) {
    background-color: ${({ theme }) =>
      lighten(0.1, theme.colorsDeprecated.primary)};

    background-image: linear-gradient(
        135deg,
        ${({ theme }) => transparentize(0.75, theme.colorsDeprecated.white)} 25%,
        transparent 25%
      ),
      linear-gradient(
        225deg,
        ${({ theme }) => transparentize(0.75, theme.colorsDeprecated.white)} 25%,
        transparent 25%
      ),
      linear-gradient(
        45deg,
        ${({ theme }) => transparentize(0.75, theme.colorsDeprecated.white)} 25%,
        transparent 25%
      ),
      linear-gradient(
        315deg,
        ${({ theme }) => transparentize(0.75, theme.colorsDeprecated.white)} 25%,
        ${({ theme }) => lighten(0.1, theme.colorsDeprecated.primary)} 25%
      );
    background-position: 10px 0, 10px 0, 0 0, 0 0;
    background-size: 10px 10px;
    background-repeat: repeat;
  }

  ${StyledBarWrapper}:nth-child(5n+3) {
    ${({ theme }) => `background: linear-gradient(-45deg, ${transparentize(
      0.9,
      theme.colorsDeprecated.white,
    )} 25%,
      ${theme.colorsDeprecated.lightViolet} 25%, ${
      theme.colorsDeprecated.lightViolet
    } 50%,
      ${transparentize(
        0.9,
        theme.colorsDeprecated.white,
      )} 50%, ${transparentize(0.9, theme.colorsDeprecated.white)} 75%, ${
      theme.colorsDeprecated.lightViolet
    }
       75%);`}
    background-size: 30px 30px;
    background-color: ${({ theme }) =>
      darken(0.2, theme.colorsDeprecated.lightViolet)};
  }

  ${StyledBarWrapper}:nth-child(5n+4) {
    background-color: ${({ theme }) => theme.colorsDeprecated.lightViolet};

    background-image: linear-gradient(
        135deg,
        ${({ theme }) => transparentize(0.8, theme.colorsDeprecated.white)} 25%,
        transparent 25%
      ),
      linear-gradient(
        225deg,
        ${({ theme }) => transparentize(0.8, theme.colorsDeprecated.white)} 25%,
        transparent 25%
      ),
      linear-gradient(
        45deg,
        ${({ theme }) => transparentize(0.8, theme.colorsDeprecated.white)} 25%,
        transparent 25%
      ),
      linear-gradient(
        315deg,
        ${({ theme }) => transparentize(0.8, theme.colorsDeprecated.white)} 25%,
        ${({ theme }) => theme.colorsDeprecated.lightViolet} 25%
      );
    background-position: 10px 0, 10px 0, 0 0, 0 0;
    background-size: 10px 10px;
    background-repeat: repeat;
  }

  ${StyledBarWrapper}:nth-child(5n+5) {
    ${({ theme }) => `background: linear-gradient(-45deg, ${transparentize(
      0.8,
      theme.colorsDeprecated.white,
    )} 25%,
      ${lighten(0.1, theme.colorsDeprecated.lightViolet)} 25%, ${lighten(
      0.1,
      theme.colorsDeprecated.lightViolet,
    )} 50%,
      ${transparentize(
        0.8,
        theme.colorsDeprecated.white,
      )} 50%, ${transparentize(
      0.8,
      theme.colorsDeprecated.white,
    )} 75%, ${lighten(0.1, theme.colorsDeprecated.lightViolet)}
       75%);`}
    background-size: 30px 30px;
    background-color: ${({ theme }) =>
      lighten(0.1, theme.colorsDeprecated.lightViolet)};
  }
`

const BarStack = ({ data, total }: BarStackProps): JSX.Element => {
  const computedTotal = useMemo(
    () => total ?? data.reduce((acc, { value }) => acc + value, 0),
    [total, data],
  )

  return (
    <StyledContainer>
      {data.map(
        ({
          id,
          value,
          text,
          onClick,
          onDoubleClick,
          onMouseEnter,
          onMouseLeave,
          onMouseDown,
          onMouseUp,
          tooltip,
        }) => (
          <StyledBarWrapper
            style={{ width: `${(value / computedTotal) * 100}%` }}
            key={id}
          >
            {tooltip ? (
              <Tooltip baseId={`tooltip-${id}`} text={tooltip}>
                <StyledBar
                  onMouseDown={onMouseDown}
                  onMouseUp={onMouseUp}
                  onClick={onClick}
                  onDoubleClick={onDoubleClick}
                  onMouseEnter={onMouseEnter}
                  onMouseLeave={onMouseLeave}
                >
                  {text}
                </StyledBar>
              </Tooltip>
            ) : (
              <StyledBar
                onMouseDown={onMouseDown}
                onMouseUp={onMouseUp}
                onClick={onClick}
                onDoubleClick={onDoubleClick}
                onMouseEnter={onMouseEnter}
                onMouseLeave={onMouseLeave}
              >
                {text}
              </StyledBar>
            )}
          </StyledBarWrapper>
        ),
      )}
    </StyledContainer>
  )
}

export default BarStack
