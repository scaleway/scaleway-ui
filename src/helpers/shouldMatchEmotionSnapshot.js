/* eslint-disable import/no-extraneous-dependencies */
import createCache from '@emotion/cache'
import { createSerializer } from '@emotion/jest'
import { CacheProvider, ThemeProvider } from '@emotion/react'
import { render } from '@testing-library/react'
import React from 'react'
import theme from '../theme'

expect.addSnapshotSerializer(createSerializer())

const emotionCache = createCache({
  key: 'cache',
})
emotionCache.compat = true

export default async (component, { transform } = {}) => {
  const node = render(
    <CacheProvider value={emotionCache}>
      <ThemeProvider theme={theme}>{component}</ThemeProvider>
    </CacheProvider>,
  )
  if (transform) await transform(node)

  expect(node.asFragment()).toMatchSnapshot()
}
