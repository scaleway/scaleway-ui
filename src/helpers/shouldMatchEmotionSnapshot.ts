/* eslint-disable import/no-extraneous-dependencies */
import { createSerializer } from '@emotion/jest'
import { RenderOptions, render } from '@testing-library/react'
import { ReactNode } from 'react'
import defaultTheme from '../theme'
import renderWithTheme from './renderWithTheme'

// use only class hash (generated from css style content)
expect.addSnapshotSerializer(
  createSerializer({ classNameReplacer: className => className }),
)

interface Options {
  options?: RenderOptions
  theme?: typeof defaultTheme
  transform?: (node: ReturnType<typeof render>) => Promise<void> | void
}

export default async (
  component: ReactNode,
  { transform, options, theme }: Options = {},
): Promise<void> => {
  const node = renderWithTheme(component, options, theme)
  if (transform) await transform(node)

  expect(node.asFragment()).toMatchSnapshot()
  node.unmount()
}
