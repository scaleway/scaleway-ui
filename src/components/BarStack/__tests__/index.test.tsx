import userEvent from '@testing-library/user-event'
import React from 'react'
import BarStack from '..'
import shouldMatchEmotionSnapshot from '../../../helpers/shouldMatchEmotionSnapshot'

const fakeData = [
  { id: '1', text: 'Hello', value: 20 },
  { id: '2', value: 42 },
  { id: '3', text: 'Encore', value: 42 },
  { id: '4', text: 'Next', value: 42 },
  { id: '5', text: 'Bye bye', value: 42 },
]

describe('BarStack', () => {
  test('should render correctly', () =>
    shouldMatchEmotionSnapshot(<BarStack data={fakeData} />))

  test('should render correctly with event handlers', async () => {
    const [
      onClick,
      onDoubleClick,
      onMouseDown,
      onMouseEnter,
      onMouseLeave,
      onMouseUp,
    ] = new Array(6).fill(1).map(() => jest.fn())

    return shouldMatchEmotionSnapshot(
      <BarStack
        data={[
          {
            id: '1',
            onClick,
            onDoubleClick,
            onMouseDown,
            onMouseEnter,
            onMouseLeave,
            onMouseUp,
            text: 'Hello',
            value: 100,
          },
        ]}
      />,
      {
        transform: node => {
          const helloDiv = node.getByText('Hello')
          userEvent.click(helloDiv)
          expect(onClick).toBeCalledTimes(1)
          expect(onMouseEnter).toBeCalledTimes(1)
          expect(onMouseDown).toBeCalledTimes(1)
          expect(onMouseUp).toBeCalledTimes(1)
          expect(onDoubleClick).toBeCalledTimes(0)
          expect(onMouseLeave).toBeCalledTimes(0)
          userEvent.dblClick(helloDiv)
          expect(onDoubleClick).toBeCalledTimes(1)
          expect(onClick).toBeCalledTimes(3)
          userEvent.unhover(helloDiv)
          expect(onMouseLeave).toBeCalledTimes(1)
        },
      },
    )
  })
})
