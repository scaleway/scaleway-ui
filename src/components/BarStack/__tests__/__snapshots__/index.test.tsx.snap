// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BarStack should render correctly 1`] = `
<DocumentFragment>
  .cache-dth6f7 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: #fff;
  border-radius: 4px;
  box-shadow: inset 0px 0px 0px 1px #d4dae7;
  overflow: hidden;
}

.cache-dth6f7 .e17nk4tl2:nth-child(5n+1) {
  background: linear-gradient(-45deg, rgba(255,255,255,0.1) 25%,
        #4f0599 25%, #4f0599 50%,
        rgba(255,255,255,0.1) 50%, rgba(255,255,255,0.1) 75%, #4f0599
         75%);
  -webkit-background-size: 30px 30px;
  background-size: 30px 30px;
  background-color: #4f0599;
}

.cache-dth6f7 .e17nk4tl2:nth-child(5n+2) {
  background-color: #6907ca;
  background-image: linear-gradient(
          135deg,
          rgba(255,255,255,0.25)
            25%,
          transparent 25%
        ),linear-gradient(
          225deg,
          rgba(255,255,255,0.25)
            25%,
          transparent 25%
        ),linear-gradient(
          45deg,
          rgba(255,255,255,0.25)
            25%,
          transparent 25%
        ),linear-gradient(
          315deg,
          rgba(255,255,255,0.25)
            25%,
          #6907ca
            25%
        );
  -webkit-background-position: 10px 0,10px 0,0 0,0 0;
  background-position: 10px 0,10px 0,0 0,0 0;
  -webkit-background-size: 10px 10px;
  background-size: 10px 10px;
  background-repeat: repeat;
}

.cache-dth6f7 .e17nk4tl2:nth-child(5n+3) {
  background: linear-gradient(-45deg, rgba(255,255,255,0.1) 25%,
        #a365f6 25%, #a365f6 50%,
        rgba(255,255,255,0.1) 50%, rgba(255,255,255,0.1) 75%, #a365f6
         75%);
  -webkit-background-size: 30px 30px;
  background-size: 30px 30px;
  background-color: #6b0ee7;
}

.cache-dth6f7 .e17nk4tl2:nth-child(5n+4) {
  background-color: #a365f6;
  background-image: linear-gradient(
          135deg,
          rgba(255,255,255,0.2)
            25%,
          transparent 25%
        ),linear-gradient(
          225deg,
          rgba(255,255,255,0.2)
            25%,
          transparent 25%
        ),linear-gradient(
          45deg,
          rgba(255,255,255,0.2)
            25%,
          transparent 25%
        ),linear-gradient(
          315deg,
          rgba(255,255,255,0.2)
            25%,
          #a365f6 25%
        );
  -webkit-background-position: 10px 0,10px 0,0 0,0 0;
  background-position: 10px 0,10px 0,0 0,0 0;
  -webkit-background-size: 10px 10px;
  background-size: 10px 10px;
  background-repeat: repeat;
}

.cache-dth6f7 .e17nk4tl2:nth-child(5n+5) {
  background: linear-gradient(-45deg, rgba(255,255,255,0.2) 25%,
        #c095f9 25%, #c095f9 50%,
        rgba(255,255,255,0.2) 50%, rgba(255,255,255,0.2) 75%, #c095f9
         75%);
  -webkit-background-size: 30px 30px;
  background-size: 30px 30px;
  background-color: #c095f9;
}

.cache-1mc9zcp {
  width: 0px;
  -webkit-transition: width 500ms;
  transition: width 500ms;
  background-color: #fff;
}

.cache-1qm6m3g {
  height: 50px;
  font-weight: 600;
  color: #fff;
  font-size: 14px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 8px;
  width: 100%;
  white-space: nowrap;
  overflow: hidden;
  text-shadow: -1px 0 rgba(74,79,98,0.3),0 1px rgba(74,79,98,0.3),1px 0 rgba(74,79,98,0.3),0 -1px rgba(74,79,98,0.3);
}

<div
    class="cache-dth6f7 e17nk4tl0"
  >
    <div
      class="cache-1mc9zcp e17nk4tl2"
      style="width: 10.638297872340425%;"
    >
      <div
        class="cache-1qm6m3g e17nk4tl1"
      >
        Hello
      </div>
    </div>
    <div
      class="cache-1mc9zcp e17nk4tl2"
      style="width: 22.340425531914892%;"
    >
      <div
        class="cache-1qm6m3g e17nk4tl1"
      />
    </div>
    <div
      class="cache-1mc9zcp e17nk4tl2"
      style="width: 22.340425531914892%;"
    >
      <div
        class="cache-1qm6m3g e17nk4tl1"
      >
        Encore
      </div>
    </div>
    <div
      class="cache-1mc9zcp e17nk4tl2"
      style="width: 22.340425531914892%;"
    >
      <div
        class="cache-1qm6m3g e17nk4tl1"
      >
        Next
      </div>
    </div>
    <div
      class="cache-1mc9zcp e17nk4tl2"
      style="width: 22.340425531914892%;"
    >
      <div
        class="cache-1qm6m3g e17nk4tl1"
      >
        Bye bye
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`BarStack should render correctly with event handlers 1`] = `
<DocumentFragment>
  .cache-dth6f7 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: #fff;
  border-radius: 4px;
  box-shadow: inset 0px 0px 0px 1px #d4dae7;
  overflow: hidden;
}

.cache-dth6f7 .e17nk4tl2:nth-child(5n+1) {
  background: linear-gradient(-45deg, rgba(255,255,255,0.1) 25%,
        #4f0599 25%, #4f0599 50%,
        rgba(255,255,255,0.1) 50%, rgba(255,255,255,0.1) 75%, #4f0599
         75%);
  -webkit-background-size: 30px 30px;
  background-size: 30px 30px;
  background-color: #4f0599;
}

.cache-dth6f7 .e17nk4tl2:nth-child(5n+2) {
  background-color: #6907ca;
  background-image: linear-gradient(
          135deg,
          rgba(255,255,255,0.25)
            25%,
          transparent 25%
        ),linear-gradient(
          225deg,
          rgba(255,255,255,0.25)
            25%,
          transparent 25%
        ),linear-gradient(
          45deg,
          rgba(255,255,255,0.25)
            25%,
          transparent 25%
        ),linear-gradient(
          315deg,
          rgba(255,255,255,0.25)
            25%,
          #6907ca
            25%
        );
  -webkit-background-position: 10px 0,10px 0,0 0,0 0;
  background-position: 10px 0,10px 0,0 0,0 0;
  -webkit-background-size: 10px 10px;
  background-size: 10px 10px;
  background-repeat: repeat;
}

.cache-dth6f7 .e17nk4tl2:nth-child(5n+3) {
  background: linear-gradient(-45deg, rgba(255,255,255,0.1) 25%,
        #a365f6 25%, #a365f6 50%,
        rgba(255,255,255,0.1) 50%, rgba(255,255,255,0.1) 75%, #a365f6
         75%);
  -webkit-background-size: 30px 30px;
  background-size: 30px 30px;
  background-color: #6b0ee7;
}

.cache-dth6f7 .e17nk4tl2:nth-child(5n+4) {
  background-color: #a365f6;
  background-image: linear-gradient(
          135deg,
          rgba(255,255,255,0.2)
            25%,
          transparent 25%
        ),linear-gradient(
          225deg,
          rgba(255,255,255,0.2)
            25%,
          transparent 25%
        ),linear-gradient(
          45deg,
          rgba(255,255,255,0.2)
            25%,
          transparent 25%
        ),linear-gradient(
          315deg,
          rgba(255,255,255,0.2)
            25%,
          #a365f6 25%
        );
  -webkit-background-position: 10px 0,10px 0,0 0,0 0;
  background-position: 10px 0,10px 0,0 0,0 0;
  -webkit-background-size: 10px 10px;
  background-size: 10px 10px;
  background-repeat: repeat;
}

.cache-dth6f7 .e17nk4tl2:nth-child(5n+5) {
  background: linear-gradient(-45deg, rgba(255,255,255,0.2) 25%,
        #c095f9 25%, #c095f9 50%,
        rgba(255,255,255,0.2) 50%, rgba(255,255,255,0.2) 75%, #c095f9
         75%);
  -webkit-background-size: 30px 30px;
  background-size: 30px 30px;
  background-color: #c095f9;
}

.cache-1mc9zcp {
  width: 0px;
  -webkit-transition: width 500ms;
  transition: width 500ms;
  background-color: #fff;
}

.cache-1qm6m3g {
  height: 50px;
  font-weight: 600;
  color: #fff;
  font-size: 14px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 8px;
  width: 100%;
  white-space: nowrap;
  overflow: hidden;
  text-shadow: -1px 0 rgba(74,79,98,0.3),0 1px rgba(74,79,98,0.3),1px 0 rgba(74,79,98,0.3),0 -1px rgba(74,79,98,0.3);
}

<div
    class="cache-dth6f7 e17nk4tl0"
  >
    <div
      class="cache-1mc9zcp e17nk4tl2"
      style="width: 100%;"
    >
      <div
        class="cache-1qm6m3g e17nk4tl1"
      >
        Hello
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`BarStack should render correctly with total 1`] = `
<DocumentFragment>
  .cache-dth6f7 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: #fff;
  border-radius: 4px;
  box-shadow: inset 0px 0px 0px 1px #d4dae7;
  overflow: hidden;
}

.cache-dth6f7 .e17nk4tl2:nth-child(5n+1) {
  background: linear-gradient(-45deg, rgba(255,255,255,0.1) 25%,
        #4f0599 25%, #4f0599 50%,
        rgba(255,255,255,0.1) 50%, rgba(255,255,255,0.1) 75%, #4f0599
         75%);
  -webkit-background-size: 30px 30px;
  background-size: 30px 30px;
  background-color: #4f0599;
}

.cache-dth6f7 .e17nk4tl2:nth-child(5n+2) {
  background-color: #6907ca;
  background-image: linear-gradient(
          135deg,
          rgba(255,255,255,0.25)
            25%,
          transparent 25%
        ),linear-gradient(
          225deg,
          rgba(255,255,255,0.25)
            25%,
          transparent 25%
        ),linear-gradient(
          45deg,
          rgba(255,255,255,0.25)
            25%,
          transparent 25%
        ),linear-gradient(
          315deg,
          rgba(255,255,255,0.25)
            25%,
          #6907ca
            25%
        );
  -webkit-background-position: 10px 0,10px 0,0 0,0 0;
  background-position: 10px 0,10px 0,0 0,0 0;
  -webkit-background-size: 10px 10px;
  background-size: 10px 10px;
  background-repeat: repeat;
}

.cache-dth6f7 .e17nk4tl2:nth-child(5n+3) {
  background: linear-gradient(-45deg, rgba(255,255,255,0.1) 25%,
        #a365f6 25%, #a365f6 50%,
        rgba(255,255,255,0.1) 50%, rgba(255,255,255,0.1) 75%, #a365f6
         75%);
  -webkit-background-size: 30px 30px;
  background-size: 30px 30px;
  background-color: #6b0ee7;
}

.cache-dth6f7 .e17nk4tl2:nth-child(5n+4) {
  background-color: #a365f6;
  background-image: linear-gradient(
          135deg,
          rgba(255,255,255,0.2)
            25%,
          transparent 25%
        ),linear-gradient(
          225deg,
          rgba(255,255,255,0.2)
            25%,
          transparent 25%
        ),linear-gradient(
          45deg,
          rgba(255,255,255,0.2)
            25%,
          transparent 25%
        ),linear-gradient(
          315deg,
          rgba(255,255,255,0.2)
            25%,
          #a365f6 25%
        );
  -webkit-background-position: 10px 0,10px 0,0 0,0 0;
  background-position: 10px 0,10px 0,0 0,0 0;
  -webkit-background-size: 10px 10px;
  background-size: 10px 10px;
  background-repeat: repeat;
}

.cache-dth6f7 .e17nk4tl2:nth-child(5n+5) {
  background: linear-gradient(-45deg, rgba(255,255,255,0.2) 25%,
        #c095f9 25%, #c095f9 50%,
        rgba(255,255,255,0.2) 50%, rgba(255,255,255,0.2) 75%, #c095f9
         75%);
  -webkit-background-size: 30px 30px;
  background-size: 30px 30px;
  background-color: #c095f9;
}

.cache-1mc9zcp {
  width: 0px;
  -webkit-transition: width 500ms;
  transition: width 500ms;
  background-color: #fff;
}

.cache-1qm6m3g {
  height: 50px;
  font-weight: 600;
  color: #fff;
  font-size: 14px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 8px;
  width: 100%;
  white-space: nowrap;
  overflow: hidden;
  text-shadow: -1px 0 rgba(74,79,98,0.3),0 1px rgba(74,79,98,0.3),1px 0 rgba(74,79,98,0.3),0 -1px rgba(74,79,98,0.3);
}

<div
    class="cache-dth6f7 e17nk4tl0"
  >
    <div
      class="cache-1mc9zcp e17nk4tl2"
      style="width: 2%;"
    >
      <div
        class="cache-1qm6m3g e17nk4tl1"
      >
        Hello
      </div>
    </div>
    <div
      class="cache-1mc9zcp e17nk4tl2"
      style="width: 4.2%;"
    >
      <div
        class="cache-1qm6m3g e17nk4tl1"
      />
    </div>
    <div
      class="cache-1mc9zcp e17nk4tl2"
      style="width: 4.2%;"
    >
      <div
        class="cache-1qm6m3g e17nk4tl1"
      >
        Encore
      </div>
    </div>
    <div
      class="cache-1mc9zcp e17nk4tl2"
      style="width: 4.2%;"
    >
      <div
        class="cache-1qm6m3g e17nk4tl1"
      >
        Next
      </div>
    </div>
    <div
      class="cache-1mc9zcp e17nk4tl2"
      style="width: 4.2%;"
    >
      <div
        class="cache-1qm6m3g e17nk4tl1"
      >
        Bye bye
      </div>
    </div>
  </div>
</DocumentFragment>
`;
